<!-- HTML header for doxygen 1.8.14-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.14"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<meta name="description" content="iVar is a computational package that contains functions broadly useful for viral amplicon-based sequencing. Additional tools for metagenomic sequencing are actively being incorporated into iVar. While each of these functions can be accomplished using existing tools, iVar contains an intersection of functionality from multiple tools that are required to call iSNVs and consensus sequences from viral sequencing data across multiple replicates. We implemented the following functions in iVar: (1) trimming of primers and low-quality bases, (2) consensus calling, (3) variant calling - both iSNVs and insertions/deletions, and (4) identifying mismatches to primer sequences and excluding the corresponding reads from alignment files."/>
<meta name="keywords" content="variants calling,consensus,htslib,primal,primer,trimming,amplicon sequencing,ivar"/>
<meta name="author" content="Andersen Lab, The Scripps Research Institute"/>
<title>iVar: Manual</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="customdoxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">iVar
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.14 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="headertitle">
<div class="title">Manual </div>  </div>
</div><!--header-->
<div class="contents">
<div class="toc"><h3>Table of Contents</h3>
<ul><li class="level1"><a href="#autotoc_md14">Available Commands</a></li>
<li class="level1"><a href="#autotoc_md15">Description of commands</a><ul><li class="level2"><a href="#autotoc_md16">Trim primer sequences with iVar</a></li>
<li class="level2"><a href="#autotoc_md17">Call variants with iVar</a></li>
<li class="level2"><a href="#autotoc_md18">Filter variants across replicates with iVar</a></li>
<li class="level2"><a href="#autotoc_md19">Generate a consensus sequences from an aligned BAM file</a></li>
<li class="level2"><a href="#autotoc_md20">Get primers with mismatches to the reference sequence</a></li>
<li class="level2"><a href="#autotoc_md21">Remove reads associated with mismatched primer indices</a></li>
<li class="level2"><a href="#autotoc_md22">(Experimental) trimadapter</a></li>
</ul>
</li>
</ul>
</div>
<div class="textblock"><h1><a class="anchor" id="autotoc_md14"></a>
Available Commands</h1>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadLeft">Command  </th><th class="markdownTableHeadLeft">Description   </th></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">trim  </td><td class="markdownTableBodyLeft">Trim reads in aligned BAM   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">variants  </td><td class="markdownTableBodyLeft">Call variants from aligned BAM file   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">filtervariants  </td><td class="markdownTableBodyLeft">Filter variants across replicates or multiple samples aligned using the same reference   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">consensus  </td><td class="markdownTableBodyLeft">Call consensus from aligned BAM file   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">getmasked  </td><td class="markdownTableBodyLeft">Detect primer mismatches and get primer indices for the amplicon to be masked   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">removereads  </td><td class="markdownTableBodyLeft">Remove reads from trimmed BAM file   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">version  </td><td class="markdownTableBodyLeft">Show version information   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">trimadapter  </td><td class="markdownTableBodyLeft">(EXPERIMENTAL) Trim adapter sequences from reads   </td></tr>
</table>
<p>To view detailed usage for each command type <code>ivar &lt;command&gt;</code> Note : Commands maked (EXPERIMENTAL) are still under active development.</p>
<h1><a class="anchor" id="autotoc_md15"></a>
Description of commands</h1>
<h2><a class="anchor" id="autotoc_md16"></a>
Trim primer sequences with iVar</h2>
<p>iVar uses primer positions supplied in a BED file to soft clip primer sequences from an aligned and sorted BAM file. Following this, the reads are trimmed based on a quality threshold(Default: 20). To do the quality trimming, iVar uses a sliding window approach(Default: 4). The windows slides from the 5' end to the 3' end and if at any point the average base quality in the window falls below the threshold, the remaining read is soft clipped. If after trimming, the length of the read is greater than the minimum length specified(Default: 30), the read is written to the new trimmed BAM file.</p>
<p>Please note that the strand is taken into account while doing the trimming so forward primers are trimmed only from forward strand and reverse primers are trimmed from reverse strand.</p>
<p>To sort and index an aligned BAM file (OPTIONAL, if index is not present iVar will create one), the following command can be used,</p>
<div class="fragment"><div class="line"># Input file - test.bam</div><div class="line">samtools sort -o test.sorted.bam test.bam &amp;&amp; samtools index test.sorted.bam.</div></div><!-- fragment --><p><b>Note</b>: All the trimming in iVar is done by soft-clipping reads in an aligned BAM file. This information is lost if reads are extracted in fastq or fasta format from the trimmed BAM file.</p>
<p>Command: </p><div class="fragment"><div class="line">ivar trim</div><div class="line"></div><div class="line">Usage: ivar trim -i &lt;input.bam&gt; -b &lt;primers.bed&gt; -p &lt;prefix&gt; [-m &lt;min-length&gt;] [-q &lt;min-quality&gt;] [-s &lt;sliding-window-width&gt;]</div><div class="line"></div><div class="line">Input Options    Description</div><div class="line">           -i    (Required) Sorted bam file, with aligned reads, to trim primers and quality</div><div class="line">           -b    (Required) BED file with primer sequences and positions</div><div class="line">           -m    Minimum length of read to retain after trimming (Default: 30)</div><div class="line">           -q    Minimum quality threshold for sliding window to pass (Default: 20)</div><div class="line">           -s    Width of sliding window (Default: 4)</div><div class="line">           -e    Include reads with no primers. By default, reads with no primers are excluded</div><div class="line"></div><div class="line">Output Options   Description</div><div class="line">           -p    (Required) Prefix for the output BAM file</div></div><!-- fragment --><p>Example Usage: </p><div class="fragment"><div class="line">ivar trim -b test_primers.bed -p test.trimmed -i test.bam -q 15 -m 50 -s 4</div></div><!-- fragment --><p>The command above will produce a trimmed BAM file test.trimmed.bam after trimming the aligned reads in test.bam using the primer positions specified in test_primers.bed and a minimum quality threshold of <b>15</b>, minimum read length of <b>50</b> and a sliding window of <b>4</b>.</p>
<p>Example BED file</p>
<div class="fragment"><div class="line">Puerto  28  52  400_1_out_L 60  +</div><div class="line">Puerto  482 504 400_1_out_R 60  -</div><div class="line">Puerto  359 381 400_2_out_L 60  +</div><div class="line">Puerto  796 818 400_2_out_R 60  -</div><div class="line">Puerto  658 680 400_3_out_L*    60  +</div><div class="line">Puerto  1054    1076    400_3_out_R*    60  -</div><div class="line">.</div><div class="line">.</div><div class="line">.</div><div class="line">.</div></div><!-- fragment --><h2><a class="anchor" id="autotoc_md17"></a>
Call variants with iVar</h2>
<p>iVar uses the output of the <code>samtools mpileup</code> command to call variants - single nucleotide variants(SNVs) and indels. In order to call variants correctly, the reference file used for alignment must be passed to iVar using the <code>-r</code> flag. The output of <code>samtools pileup</code> is piped into <code>ivar variants</code> to generate a .tsv file with the variants. There are two parameters that can be set for variant calling using iVar - minimum quality(Default: 20) and minimum frequency(Default: 0.03). Minimum quality is the minimum quality for a base to be counted towards the ungapped depth to canculate iSNV frequency at a given position. For insertions, the quality metric is discarded and the mpileup depth is used directly. Minimum frequency is the minimum frequency required for a SNV or indel to be reported.</p>
<h4>Amino acid translation of iSNVs</h4>
<p>iVar can identify codons and translate variants into amino acids using a GFF file in the https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md "GFF3" format containing the required coding regions (CDS). In absence of a GFF file, iVar will not perform the translation and "NA" will be added to the output file in place of the reference and alternate codons and amino acids. The GFF file in the GFF3 format can be downloaded via ftp from NCBI RefSeq/Genbank. They are usually the files with the extension ".gff.gz". For example, the GFF file for Zaire Ebolavirus can be found <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/viral/Zaire_ebolavirus/all_assembly_versions/GCF_000848505.1_ViralProj14703">here</a>. More details on GFF3 files hosted by NCBI can be found in their ftp <a href="https://www.ncbi.nlm.nih.gov/genome/doc/ftpfaq/">FAQs</a>.</p>
<h4>Account for RNA editing through polymerase slippage</h4>
<p>Some RNA viruses such as Ebola virus, might have polymerase slippage causing the insertion of a couple of nucleotides. More details can be found <a href="https://viralzone.expasy.org/857?outline=all_by_protein">here</a>. iVar can account for this editing and identify the correct open reading frames. The user will have to specify two additional parameters, <b>EditPosition</b>: Position at which edit occurs and <b>EditSequence</b>: The sequence tht is inserted at the given positon, in the "attributes" column of the GFF file to account for this. A test example is given below,</p>
<div class="fragment"><div class="line">test    Genbank CDS 2   292 .   +   .   ID=id-testedit1;Note=PinkFloyd;EditPosition=100;EditSequence=A</div><div class="line">test    Genbank CDS 2   292 .   +   .   ID=id-testedit2;Note=AnotherBrickInTheWall;EditPosition=102;EditSequence=AA</div></div><!-- fragment --><p>If a certain base is present in multiple CDSs, iVar will add a new row for each CDS frame and distinguish the rows by adding the ID (specified in attributes of GFF) of the GFF feature used for the translation. This is shown for position 42 in the example output below. There are two rows with two different GFF features: id-test3 and id-test4.</p>
<p>Command: </p><div class="fragment"><div class="line">Usage: samtools mpileup -aa -A -d 0 -B -Q 0 --reference [&lt;reference-fasta] &lt;input.bam&gt; | ivar variants -p &lt;prefix&gt; [-q &lt;min-quality&gt;] [-t &lt;min-frequency-threshold&gt;] [-m &lt;minimum depth&gt;] [-r &lt;reference-fasta&gt;] [-g GFF file]</div><div class="line"></div><div class="line">Note : samtools mpileup output must be piped into ivar variants</div><div class="line"></div><div class="line">Input Options    Description</div><div class="line">           -q    Minimum quality score threshold to count base (Default: 20)</div><div class="line">           -t    Minimum frequency threshold(0 - 1) to call variants (Default: 0.03)</div><div class="line">           -m    Minimum read depth to call variants (Default: 0)</div><div class="line">           -r    Reference file used for alignment. This is used to translate the nucleotide sequences and identify intra host single nucleotide variants</div><div class="line">           -g    A GFF file in the GFF3 format can be supplied to specify coordinates of open reading frames (ORFs). In absence of GFF file, amino acid translation will not be done.</div><div class="line"></div><div class="line">Output Options   Description</div><div class="line">           -p    (Required) Prefix for the output tsv variant file</div></div><!-- fragment --><p>Example Usage: </p><div class="fragment"><div class="line">samtools mpileup -aa -A -d 600000 -B -Q 0 test.trimmed.bam | ivar variants -p test -q 20 -t 0.03 -r test_reference.fa -g test.gff</div></div><!-- fragment --><p>The command above will generate a test.tsv file.</p>
<p>Example of output .tsv file.</p>
<div class="fragment"><div class="line">REGION  POS REF ALT REF_DP  REF_RV  REF_QUAL    ALT_DP  ALT_RV  ALT_QUAL    ALT_FREQ    TOTAL_DP    PVAL    PASS    GFF_FEATURE REF_CODON   REF_AA  ALT_CODON   ALT_AA</div><div class="line">test    42  G   T   0   0   0   1   0   49  1   1   1   FALSE   id-test3    AGG R   ATG M</div><div class="line">test    42  G   T   0   0   0   1   0   49  1   1   1   FALSE   id-test4    CAG Q   CAT H</div><div class="line">test    320 A   T   1   1   35  1   1   46  0.5 2   0.666667    FALSE   NA  NA  NA  NA  NA</div><div class="line">test    365 A   T   0   0   0   1   1   27  1   1   1   FALSE   NA  NA  NA  NA  NA</div></div><!-- fragment --><p>Description</p>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadLeft">Field  </th><th class="markdownTableHeadLeft">Description   </th></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">REGION  </td><td class="markdownTableBodyLeft">Region from BAM file   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">POS  </td><td class="markdownTableBodyLeft">Position on reference sequence   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">REF  </td><td class="markdownTableBodyLeft">Reference base   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">ALT  </td><td class="markdownTableBodyLeft">Alternate Base   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">REF_DP  </td><td class="markdownTableBodyLeft">Ungapped depth of reference base   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">REF_RV  </td><td class="markdownTableBodyLeft">Ungapped depth of reference base on reverse reads   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">REF_QUAL  </td><td class="markdownTableBodyLeft">Mean quality of reference base   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">ALT_DP  </td><td class="markdownTableBodyLeft">Ungapped depth of alternate base.   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">ALT_RV  </td><td class="markdownTableBodyLeft">Ungapped deapth of alternate base on reverse reads   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">ALT_QUAL  </td><td class="markdownTableBodyLeft">Mean quality of alternate base   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">ALT_FREQ  </td><td class="markdownTableBodyLeft">Frequency of alternate base   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">TOTAL_DP  </td><td class="markdownTableBodyLeft">Total depth at position   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">PVAL  </td><td class="markdownTableBodyLeft">p-value of fisher's exact test   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">PASS  </td><td class="markdownTableBodyLeft">Result of p-value &lt;= 0.05   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">GFF_FEATURE  </td><td class="markdownTableBodyLeft">ID of the GFF feature used for the translation   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">REF_CODON  </td><td class="markdownTableBodyLeft">Codong using the reference base   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">REF_AA  </td><td class="markdownTableBodyLeft">Amino acid translated from reference codon   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">ALT_CODON  </td><td class="markdownTableBodyLeft">Codon using the alternate base   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">ALT_AA  </td><td class="markdownTableBodyLeft">Amino acid translated from the alternate codon   </td></tr>
</table>
<p><b>Note</b>: Please use the -B options with <code>samtools mpileup</code> to call variants and generate consensus. When a reference sequence is supplied, the quality of the reference base is reduced to 0 (ASCII: !) in the mpileup output. Disabling BAQ with -B seems to fix this. This was tested in samtools 1.7 and 1.8.</p>
<h2><a class="anchor" id="autotoc_md18"></a>
Filter variants across replicates with iVar</h2>
<p>iVar can be used to get an intersection of variants(in .tsv files) called from any number of replicates or from different samples using the same reference sequence. This intersection will filter out any iSNVs that do not occur in a minimum fraction of the files supplied. This parameter can be changed using the <code>-t</code> flag which range from 0 to 1 (default). Fields that are different across replicates(fields apart from REGION, POS, REF, ALT, REF_CODON, REF_AA, ALT_CODON, ALT_AA) will have the filename added as a suffix. If there are a large number of files to be filtered, the <code>-f</code> flag can be used to supply a text file with one sample/replicate variant .tsv file per line.</p>
<p>Command: </p><div class="fragment"><div class="line">Usage: ivar filtervariants -p &lt;prefix&gt; replicate-one.tsv replicate-two.tsv ... OR ivar filtervariants -p &lt;prefix&gt; -f &lt;text file with one variant file per line&gt; </div><div class="line">Input: Variant tsv files for each replicate/sample</div><div class="line"></div><div class="line">Input Options    Description</div><div class="line">           -t    Minimum fration of files required to contain the same variant. Specify value within [0,1]. (Default: 1)</div><div class="line">           -f    A text file with one variant file per line.</div><div class="line"></div><div class="line">Output Options   Description</div><div class="line">           -p    (Required) Prefix for the output filtered tsv file</div></div><!-- fragment --><p>Example Usage: The command below only retains those variants that are found in atleast 50% of the fiels supplied </p><div class="fragment"><div class="line">ivar filtervariants -t 0.5 -p test.filtered test.1.tsv test.2.tsv test.3.tsv</div></div><!-- fragment --><p>The three replicates can also be supplied using a text file as shown below</p>
<div class="fragment"><div class="line">ivar filtervariants -t 0.5 -p test.filtered -f filter_files.txt</div></div><!-- fragment --><p>filter_files.txt </p><div class="fragment"><div class="line">./path/to/test.1.tsv</div><div class="line">./path/to/test.2.tsv</div><div class="line">./path/to/test.3.tsv</div></div><!-- fragment --><p>The command above will prodoce an output .tsv file test.filtered.tsv.</p>
<p>Example output of filtered .tsv file from three files test_rep1.tsv and test_rep2.tsv</p>
<div class="fragment"><div class="line">REGION  POS REF ALT GFF_FEATURE REF_CODON   REF_AA  ALT_CODON   ALT_AA  REF_DP_test.1.tsv   REF_RV_test.1.tsv   REF_QUAL_test.1.tsv ALT_DP_test.1.tsv   ALT_RV_test.1.tsv   ALT_QUAL_test.1.tsv ALT_FREQ_test.1.tsv TOTAL_DP_test.1.tsv PVAL_test.1.tsv PASS_test.1.tsv REF_DP_test.2.tsv   REF_RV_test.2.tsv   REF_QUAL_test.2.tsv ALT_DP_test.2.tsv   ALT_RV_test.2.tsv   ALT_QUAL_test.2.tsv ALT_FREQ_test.2.tsv TOTAL_DP_test.2.tsv PVAL_test.2.tsv PASS_test.2.tsv REF_DP_test.3.tsv   REF_RV_test.3.tsv   REF_QUAL_test.3.tsv ALT_DP_test.3.tsv   ALT_RV_test.3.tsv   ALT_QUAL_test.3.tsv ALT_FREQ_test.3.tsv TOTAL_DP_test.3.tsv PVAL_test.3.tsv PASS_test.3.tsv </div><div class="line">test    139 T   A   id-test3    GCT A   GCA A   1   0   32  1   0   55  0.5 2   0.666667    FALSE   1   0   32  1   0   55  0.5 2   0.666667    FALSE   NA  NA  NA  NA  NA  NA  NA  NA  NA  NA</div><div class="line">test    320 A   T   NA  NA  NA  NA  NA  1   1   35  1   1   46  0.5 2   0.666667    FALSE   NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  1   1   35  1   1   46  0.5 2   0.666667    FALSE</div><div class="line">test    365 A   T   NA  NA  NA  NA  NA  0   0   0   1   1   27  1   1   1   FALSE   0   0   0   1   1   27  1   1   1   FALSE   0   0   0   1   1   27  1   1   1   FALSE</div><div class="line">test    42  G   T   id-test4    CAG Q   CAT H   0   0   0   1   0   49  1   1   1   FALSE   0   0   0   1   0   49  1   1   1   FALSE   NA  NA  NA  NA  NA  NA  NA  NA  NA  NA</div><div class="line">test    42  G   T   id-testedit1    AGG R   ATG M   0   0   0   1   0   49  1   1   1   FALSE   0   0   0   1   0   49  1   1   1   FALSE   0   0   0   1   0   49  1   1   1   FALSE</div><div class="line">test    69  T   G   id-testedit2    TTG L   TGG W   1   0   57  1   0   53  0.5 2   0.666667    FALSE   1   0   57  1   0   53  0.5 2   0.666667    FALSE   1   0   57  1   0   53  0.5 2   0.666667    FALSE</div></div><!-- fragment --><p>Description of fields</p>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadLeft">No  </th><th class="markdownTableHeadLeft">Field  </th><th class="markdownTableHeadNone">Description   </th></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">1  </td><td class="markdownTableBodyLeft">REGION  </td><td class="markdownTableBodyNone">Common region across all replicate variant tsv files   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">2  </td><td class="markdownTableBodyLeft">POS  </td><td class="markdownTableBodyNone">Common position across all variant tsv files   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">3  </td><td class="markdownTableBodyLeft">REF  </td><td class="markdownTableBodyNone">Common reference base across all variant tsv files   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">4  </td><td class="markdownTableBodyLeft">ALT  </td><td class="markdownTableBodyNone">Common alternate base across all variant tsv files   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">5  </td><td class="markdownTableBodyLeft">GFF_FEATURE  </td><td class="markdownTableBodyNone">GFF feature used for the translation   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">6  </td><td class="markdownTableBodyLeft">REF_CODON  </td><td class="markdownTableBodyNone">The codon using the reference base   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">7  </td><td class="markdownTableBodyLeft">REF_AA  </td><td class="markdownTableBodyNone">Reference codon translated into amino acid   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">8  </td><td class="markdownTableBodyLeft">ALT_CODON  </td><td class="markdownTableBodyNone">Codon using the alternate base   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">9  </td><td class="markdownTableBodyLeft">ALT_AA  </td><td class="markdownTableBodyNone">Alternate codon translated into amino acid   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">10  </td><td class="markdownTableBodyLeft">REF_DP_&lt;rep1-tsv-file-name&gt;  </td><td class="markdownTableBodyNone">Depth of reference base in replicate 1   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">11  </td><td class="markdownTableBodyLeft">REF_RV_&lt;rep1-tsv-file-name&gt;  </td><td class="markdownTableBodyNone">Depth of reference base on reverse reads in replicate 1   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">12  </td><td class="markdownTableBodyLeft">REF_QUAL_&lt;rep1-tsv-file-name&gt;  </td><td class="markdownTableBodyNone">Mean quality of reference base in replicate 1   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">13  </td><td class="markdownTableBodyLeft">ALT_DP_&lt;rep1-tsv-file-name&gt;  </td><td class="markdownTableBodyNone">Depth of alternate base in replicate 1   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">14  </td><td class="markdownTableBodyLeft">ALT_RV_&lt;rep1-tsv-file-name&gt;  </td><td class="markdownTableBodyNone">Deapth of alternate base on reverse reads in replicate 1   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">15  </td><td class="markdownTableBodyLeft">ALT_QUAL_&lt;rep1-tsv-file-name&gt;  </td><td class="markdownTableBodyNone">Mean quality of alternate base in replicate 1   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">16  </td><td class="markdownTableBodyLeft">ALT_FREQ_&lt;rep1-tsv-file-name&gt;  </td><td class="markdownTableBodyNone">Frequency of alternate base in replicate 1   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">17  </td><td class="markdownTableBodyLeft">TOTAL_DP_&lt;rep1-tsv-file-name&gt;  </td><td class="markdownTableBodyNone">Total depth at position in replicate 1   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">18  </td><td class="markdownTableBodyLeft">PVAL_&lt;rep1-tsv-file-name&gt;  </td><td class="markdownTableBodyNone">p-value of fisher's exact test in replicate 1   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">19  </td><td class="markdownTableBodyLeft">PASS_&lt;rep1-tsv-file-name&gt;  </td><td class="markdownTableBodyNone">Result of p-value &lt;= 0.05 in replicate 1   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">20  </td><td class="markdownTableBodyLeft">Continue rows 10 - 19 for every replicate provided  </td><td class="markdownTableBodyNone"></td></tr>
</table>
<h2><a class="anchor" id="autotoc_md19"></a>
Generate a consensus sequences from an aligned BAM file</h2>
<p>To generate a consensus sequence iVar uses the output of <code>samtools mpileup</code> command. The mpileup output must be piped into <code>ivar consensus</code>. There are five parameters that can be set - minimum quality(Default: 20), minimum frequency threshold(Default: 0), minimum depth to call a consensus(Default: 10), a flag to exclude nucleotides from regions with depth less than the minimum depth and a character to call in regions with coverage lower than the speicifed minimum depth(Default: 'N'). Minimum quality is the minimum quality of a base to be considered in calculations of variant frequencies at a given position. Minimum frequency threshold is the minimum frequency that a base must match to be called as the consensus base at a position. If one base is not enough to match a given frequency, then an ambigious nucleotide is called at that position. Minimum depth is the minimum required depth to call a consensus. If '-k' flag is set then these regions are not included in the consensus sequence. If '-k' is not set then by default, a 'N' is called in these regions. You can also specfy which character you want to add to the consensus to cover regions with depth less than the minimum depth. This can be done using -n option. It takes one of two values: '-' or 'N'.</p>
<p>As an example, consider a position with 6As, 3Ts and 1C. The table below shows the consensus nucleotide called at different frequencies.</p>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadLeft">Minimum frequency threshold  </th><th class="markdownTableHeadNone">Consensus   </th></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">0  </td><td class="markdownTableBodyNone">A   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">0.5  </td><td class="markdownTableBodyNone">A   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">0.6  </td><td class="markdownTableBodyNone">A   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">0.7  </td><td class="markdownTableBodyNone">W(A or T)   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">0.9  </td><td class="markdownTableBodyNone">W (A or T)   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">1  </td><td class="markdownTableBodyNone">H (A or T or C)   </td></tr>
</table>
<p>If there are two nucleotides at the same frequency, both nucleotides are used to call an ambigious base as the consensus. As an example, consider a position wiht 6 Ts, 2As and 2 Gs. The table below shows the consensus nucleotide called at different frequencies.</p>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadLeft">Minimum frequency threshold  </th><th class="markdownTableHeadNone">Consensus   </th></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">0  </td><td class="markdownTableBodyNone">T   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">0.5  </td><td class="markdownTableBodyNone">T   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">0.6  </td><td class="markdownTableBodyNone">T   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">0.7  </td><td class="markdownTableBodyNone">D(A or T or G)   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">0.9  </td><td class="markdownTableBodyNone">D(A or T or G)   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">1  </td><td class="markdownTableBodyNone">D(A or T or G)   </td></tr>
</table>
<p>The output of the command is a fasta file with the consensus sequence and a .txt file with the average quality of every base used to generate the consensus at each position. <em>For insertions, the quality is set to be the minimum quality threshold since mpileup doesn't give the quality of bases in insertions.</em></p>
<p>Command:</p>
<div class="fragment"><div class="line">ivar consensus</div><div class="line"></div><div class="line">Usage: samtools mpileup -aa -A -d 0 -Q 0 &lt;input.bam&gt; | ivar consensus -p &lt;prefix&gt; </div><div class="line"></div><div class="line">Note : samtools mpileup output must be piped into ivar consensus</div><div class="line"></div><div class="line">Input Options    Description</div><div class="line">           -q    Minimum quality score threshold to count base (Default: 20)</div><div class="line">           -t    Minimum frequency threshold(0 - 1) to call consensus. (Default: 0)</div><div class="line">                 Frequently used thresholds | Description</div><div class="line">                 ---------------------------|------------</div><div class="line">                                          0 | Majority or most common base</div><div class="line">                                        0.2 | Bases that make up atleast 20% of the depth at a position</div><div class="line">                                        0.5 | Strict or bases that make up atleast 50% of the depth at a position</div><div class="line">                                        0.9 | Strict or bases that make up atleast 90% of the depth at a position</div><div class="line">                                          1 | Identical or bases that make up 100% of the depth at a position. Will have highest ambiguities</div><div class="line">           -m    Minimum depth to call consensus(Default: 10)</div><div class="line">           -k    If &#39;-k&#39; flag is added, regions with depth less than minimum depth will not be added to the consensus sequence. Using &#39;-k&#39; will override any option specified using -n </div><div class="line">           -n    (N/-) Character to print in regions with less than minimum coverage(Default: N)</div><div class="line"></div><div class="line">Output Options   Description</div><div class="line">           -p    (Required) Prefix for the output fasta file and quality file</div></div><!-- fragment --><p>Example Usage: </p><div class="fragment"><div class="line">samtools mpileup -d 1000 -A -Q 0 test.bam | ivar consensus -p test -q 20 -t 0</div></div><!-- fragment --><p>The command above will produce a test.fa fasta file with the consensus sequence and a test.qual.txt with the average quality of each base in the consensus sequence.</p>
<h2><a class="anchor" id="autotoc_md20"></a>
Get primers with mismatches to the reference sequence</h2>
<p>iVar uses a .tsv file with variants to get the zero based indices(based on the BED file) of mismatched primers. This command requires another .tsv file with each line containing the left and right primer names separated by a tab. This is used to get both the primers for an amplicon with a single mismatched primer. The output is a text file with the zero based primer indices delimited by a space. The output is written to a a text file using the prefix provided.</p>
<p>Command: </p><div class="fragment"><div class="line">ivar getmasked</div><div class="line">Usage: ivar getmasked -i &lt;input-filtered.tsv&gt; -b &lt;primers.bed&gt; -f &lt;primer_pairs.tsv&gt; -p &lt;prefix&gt;</div><div class="line">Note: This step is used only for amplicon-based sequencing.</div><div class="line"></div><div class="line">Input Options    Description</div><div class="line">           -i    (Required) Input filtered variants tsv generated from &#39;ivar filtervariants&#39;</div><div class="line">           -b    (Required) BED file with primer sequences and positions</div><div class="line">           -f    (Required) Primer pair information file containing left and right primer names for the same amplicon separated by a tab</div><div class="line">Output Options   Description</div><div class="line">           -p    (Required) Prefix for the output text file</div></div><!-- fragment --><p>Example BED file</p>
<div class="fragment"><div class="line">Puerto  28  52  400_1_out_L 60  +</div><div class="line">Puerto  482 504 400_1_out_R 60  -</div><div class="line">Puerto  359 381 400_2_out_L 60  +</div><div class="line">Puerto  796 818 400_2_out_R 60  -</div><div class="line">Puerto  658 680 400_3_out_L*    60  +</div><div class="line">Puerto  1054    1076    400_3_out_R*    60  -</div><div class="line">.</div><div class="line">.</div><div class="line">.</div><div class="line">.</div></div><!-- fragment --><p>Example primer pair information file </p><div class="fragment"><div class="line">400_1_out_L    400_1_out_R</div><div class="line">400_2_out_L    400_2_out_R</div><div class="line">400_3_out_L    400_3_out_R</div><div class="line">.</div><div class="line">.</div><div class="line">.</div><div class="line">.</div></div><!-- fragment --><p>Example Usage: </p><div class="fragment"><div class="line">ivar getmasked -i test.filtered.tsv -b primers.bed -f pair_information.tsv -p test.masked.txt</div></div><!-- fragment --><p>The command above produces an output file - test.masked.txt.</p>
<p>Example Output:</p>
<div class="fragment"><div class="line">1 2 7 8</div></div><!-- fragment --><h2><a class="anchor" id="autotoc_md21"></a>
Remove reads associated with mismatched primer indices</h2>
<p>This command accepts an aligned and sorted BAM file trimmed using <code>ivar trim</code> and removes the reads corresponding to the supplied primer indices, which is the output of <code>ivar getmasked</code> command. Under the hood, <code>ivar trim</code> adds the zero based primer index(based on the BED file) to the BAM auxillary data for every read. Hence, ivar removereads will only work on BAM files that have been trimmed using <code>ivar trim</code>.</p>
<p>Command: </p><div class="fragment"><div class="line">ivar removereads</div><div class="line"></div><div class="line">Usage: ivar removereads -i &lt;input.trimmed.bam&gt; -p &lt;prefix&gt; -t &lt;text-file-with-primer-indices&gt;</div><div class="line">Note: This step is used only for amplicon-based sequencing.</div><div class="line"></div><div class="line">Input Options    Description</div><div class="line">           -i    (Required) Input BAM file  trimmed with ivar trim. Must be sorted and indexed, which can be done using sort_index_bam.sh</div><div class="line">           -t    (Required) Text file with primer indices separated by spaces. This is the output of getmasked command.</div><div class="line"></div><div class="line">Output Options   Description</div><div class="line">           -p    (Required) Prefix for the output filtered BAM file</div></div><!-- fragment --><p>Example Usage: </p><div class="fragment"><div class="line">ivar trim -i test.bam -p test.trimmed</div><div class="line">ivar removereads -i test.trimmed.bam -p test.trimmed.masked.bam -t test.masked.txt</div></div><!-- fragment --><p>The <code>ivar trim</code> command above trims test.bam and produced test.trimmed.bam with the primer indice data added. The <code>ivar removereads</code> command produces an output file - test.trimmed.masked.bam after removing all the reads corresponding to primer indices - 1,2,7 and 8.</p>
<h2><a class="anchor" id="autotoc_md22"></a>
(Experimental) trimadapter</h2>
<p><b>Note: This feature is under active development and not completely validated yet.</b></p>
<p>trimadapter in iVar can be used to trim adapter sequences from fastq files using a supplied fasta file. </p>
</div></div><!-- contents -->
<!-- HTML footer for doxygen 1.8.14-->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.14
</small></address>
</body>
</html>
